### PR TITLE
feat(nudging): optional specific-humidity relaxation, and a public vertical_interp_log_p

### PR DESCRIPTION
**Pull request by JOhnsonKC201**
## Describe your changes

The API slice of #764. That PR was closed — the NN bias correction itself now lives in [`jcm-experiments/nn-bias-correction`](https://github.com/climate-analytics-lab/jcm-experiments/tree/main/nn-bias-correction) — but two of the changes in it are in the main API and are worth having on their own. This is those two and nothing else: 4 files, +224/−27.

**Optional specific-humidity relaxation.** Nudging could relax winds and temperature but not humidity. `NudgingTarget` gains `specific_humidity` and `NudgingConfig` gains `inv_tau_humidity`. Both default to `None`, meaning not relaxed, and both are positioned last, so every existing target and config keeps working with the argument count it has today. `nudging_tendency` returns a zero humidity tendency unless both are supplied, and it decides that with a Python-level `is None` rather than a traced `jnp.where`, so jit resolves it at trace time and the wind-and-temperature path costs exactly what it did before. `from_dataset` gains `q_var` and leaves `specific_humidity` as `None` when the variable is absent, so a dataset without humidity never reads as a dry atmosphere, and `NudgingConfig.temp_humidity` builds the T/q profile — the complement of the `winds_only` that was already there.

The humidity in a `NudgingTarget` is in g/kg, the `PhysicsState` convention, and the docstring says so. Raw ERA5 is kg/kg, so getting this wrong nudges the model toward one thousandth of the intended value and nothing anywhere raises. That is the whole reason for the docs hunk below.

**`vertical_interp_log_p` made public.** Nothing in the ozone climatology's log-pressure interpolation is ozone-specific beyond the argument names, and an ERA5 loader outside this repository reuses it for temperature, humidity and winds. Importing an underscore name across packages couples to a private detail. There are exactly two references to it in the whole repository, both inside `interpolate_ozone.py`; the internal call site follows the rename.

**A units note in `getting_started.rst`,** five lines next to the nudging example, saying what units the target fields are in.

## Testing

Branched off current `dev` (`f96575c1`). Neither `nudging.py` nor `interpolate_ozone.py` has been touched on `dev` since I first wrote these, so they apply cleanly.

`pytest jcm/nudging_test.py` gives 21 passed, 3 subtests passed. Eight of those cases are new or changed: `winds_only` zeroing humidity, `temp_humidity` zeroing winds, the sign of the humidity tendency, humidity absent (static and time-varying, both staying `None`) and present in `from_dataset`, the humidity tendency staying exactly zero when a target without `q` meets `temp_humidity`, and a grid-versus-flattened-column agreement test that pins the broadcasting down rather than leaving it asserted in a docstring. `pytest jcm/data/bc/` gives 9 passed. `ruff check` with the pinned 0.15.17 is clean. All of that in a fresh CPU environment built from `requirements.txt` plus the forked dinosaur.

The `claude-review` check fails with an empty `ANTHROPIC_API_KEY` in its log — fork PRs do not get repository secrets, so I do not think that one is about this branch.

## Issue ticket number and link

#356, and the closed #764.

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by duncanwp** [comment]
@codex review

**Comment by duncanwp** [comment]
@codex review

File: jcm/nudging.py [hunk 1]
```diff
@@ -99,15 +112,19 @@ def to_jax(name):
         u = to_jax(u_var)
         v = to_jax(v_var)
         T = to_jax(T_var)
+        # Humidity is optional: a wind/temperature-only target (the original
+        # use case) carries no q field, so fall back to zeros.
+        q = to_jax(q_var) if q_var in ds else jnp.zeros_like(T)
```
[the remark this answers is not in the document]
Comment (JOhnsonKC201) [reply jcm/nudging.py]: Confirmed, and fixed in 4605b772.

This was worse than the docstring admitted. `None` is the documented "no
humidity reference" sentinel and `nudging_tendency` already branches on it,
but `from_dataset` could never produce one: it substituted `jnp.zeros_like(T)`.
Zeros are not an absent reference, they are a reference to a bone-dry
atmosphere, so the sentinel branch was unreachable from the public
constructor.

You are also right about the ERA5 path, and that is where it actually bites.
`nudging_target()` called `dataset_on_model_grid()` with its default
`variables=("u", "v", "T")`, so the repository's standard target carried no
`q`. Paired with `NudgingConfig.temp_humidity()`, which sets a non-zero
`inv_tau_humidity` on every level, that is precisely the supervised
bias-correction setup this PR exists to enable, and it would have been
trained against a model being relaxed toward zero humidity.

Two things were hiding it: `test_humidity_absent_is_zero` asserted the zero
fill outright, and the tendency direction test built a q-less target and read
the resulting drying as expected, commented `# state q > target q (0)`.

The fix:

- `from_dataset` keeps `specific_humidity=None` when `q` is absent, on the
  static and the `TimeSeries` path both.
- `era5.nudging_target` requests `q` by default, with `humidity=False` for
  callers who genuinely want a winds-only target and would rather not pay for
  the extra variable.
- The zero-fill test is replaced by one pinning the `None` contract and one
  asserting the tendency is exactly zero under `temp_humidity`; the direction
  test now supplies a real humidity reference so it tests direction rather
  than the fallback.

File: jcm/nudging.py [hunk 2]
```diff
@@ -276,12 +326,22 @@ def _match(ref, like):
...
+    # matches the wind/temperature-only behaviour this module had before
+    # humidity became configurable. Both are checked on the Python object
+    # rather than traced: each is either present for the whole run or absent
+    # for the whole run.
+    if target.specific_humidity is None or config.inv_tau_humidity is None:
+        q_t = jnp.zeros_like(state.specific_humidity)
+    else:
+        q_t = inv_tau_q * (
+            _match(target.specific_humidity, state.specific_humidity)
+            - state.specific_humidity)
```
[the remark this answers is not in the document]
Comment (JOhnsonKC201) [reply jcm/nudging.py:343]: The unit mismatch on Pyses is real, but I do not think the conclusion follows
that the target convention here is the thing to change. The g/kg convention is
not something this PR invented; it is the documented repository-wide
`PhysicsState` contract, and `PysesCamSEDycore` is the component that does not
honour it.

Evidence:

- `jcm/dynamics_units_table.csv:4` is the canonical table:
  `specific_humidity,g/kg,qa,Specific humidity`.
- `jcm/dycore/dinosaur/state_bridge.py:124` builds the `PhysicsState` with
  `dimensionalize(q, units.gram / units.kilogram)`.
- `jcm/diagnostics.py:110-119` reads the saved `specific_humidity` as g/kg,
  applies a `q_max > 100 g/kg` health threshold, and carries a comment about
  an earlier version that assumed kg/kg and applied a spurious `* 1000`.
- `jcm/initial_states/injectors.py:181-186` documents this exact trap for
  injection: raw kg/kg made the physics see `q` 1000x too large.

Against that, `jcm/dycore/pyses/dycore.py:742-743` builds its `PhysicsState`
with `q = r_pos / (1.0 + r_pos)`, the dimensionless kg/kg fraction, and
lines 819-821 consume `tend.specific_humidity` in those same kg/kg units. So
a Pyses `PhysicsState` is 1000x below the value every other consumer of that
field assumes, which is a problem for the physics parameterisations and the
diagnostics thresholds independently of whether nudging is switched on.

That makes it a pre-existing Pyses conformance bug rather than something this
PR introduces or should carry. This PR changes `docs/source/getting_started.rst`,
`jcm/data/bc/interpolate_ozone.py`, `jcm/nudging.py` and `jcm/nudging_test.py`,
and touches no Pyses file. Correcting the Pyses humidity convention means
changing the state bridge and the tendency scatter together, and revalidating
every Pyses physics run against the change, which wants its own PR and its own
review.

@duncanwp happy to open a separate issue against the Pyses backend with the
above if you agree with the reading. If you would rather nudging refuse to run
on a backend whose humidity units it cannot vouch for, I can add that guard
here instead, but converting targets per backend inside `nudging.py` would
bake the inconsistency in rather than surface it.

The second P1 on this review was a genuine bug in this PR and is fixed in
4605b772.

**Comment by duncanwp** [comment]
@codex review

File: jcm/data/era5.py [hunk 3]
```diff
@@ -256,10 +256,21 @@ def dataset_on_model_grid(coords, start: str, end: str, *,
...
+    sets one on every level) is a configuration error rather than a
+    humidity-free run. Pass ``humidity=False`` for a winds-only target to
+    skip downloading and caching the extra variable.
+    """
     from jcm.nudging import NudgingTarget
-    ds = dataset_on_model_grid(coords, start, end, freq=freq, cache=cache)
+    variables = ("u", "v", "T", "q") if humidity else ("u", "v", "T")
+    ds = dataset_on_model_grid(coords, start, end, freq=freq,
+                               variables=variables, cache=cache)
     return NudgingTarget.from_dataset(ds)
```
[the remark this answers is not in the document]
Comment (JOhnsonKC201) [reply jcm/data/era5.py:263]: Agreed that it was broken. I have taken the `era5.py` change out of this PR instead of fixing it here, in 59ee591. `nudging_target` is back to fetching u, v and T only, so it never passes raw ERA5 `q` into a target. Without `q` the target's humidity is `None`, and `nudging_tendency` gives a zero humidity tendency, so nothing gets relaxed toward kg/kg values.

Fetching `q` by default was a new feature, not part of the fix for the previous thread, and doing it properly needs the x1000 conversion and a test for it. If anyone wants ERA5 humidity through `jcm.data.era5`, I can do that as a separate PR.

File: jcm/data/era5.py [hunk 4]
```diff
@@ -256,10 +256,21 @@ def dataset_on_model_grid(coords, start: str, end: str, *,
...
+    ``inv_tau_humidity`` (:meth:`~jcm.nudging.NudgingConfig.temp_humidity`
+    sets one on every level) is a configuration error rather than a
+    humidity-free run. Pass ``humidity=False`` for a winds-only target to
+    skip downloading and caching the extra variable.
+    """
     from jcm.nudging import NudgingTarget
-    ds = dataset_on_model_grid(coords, start, end, freq=freq, cache=cache)
+    variables = ("u", "v", "T", "q") if humidity else ("u", "v", "T")
+    ds = dataset_on_model_grid(coords, start, end, freq=freq,
+                               variables=variables, cache=cache)
```
[the remark this answers is not in the document]
Comment (JOhnsonKC201) [reply jcm/data/era5.py]: Same change, 59ee591: `nudging_target` no longer asks for `q`, so it requests the same `("u", "v", "T")` as the `_main` prefetch again and the cache key matches.

File: jcm/nudging.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JOhnsonKC201) [reply jcm/nudging.py]: Update: I've since dropped the `era5.nudging_target` part of this in 59ee591 (see the two threads on `era5.py`). The `from_dataset` fix and its tests stay.

**Comment by duncanwp** [comment]
@codex review

File: jcm/nudging.py [hunk 5]
```diff
@@ -276,12 +330,22 @@ def _match(ref, like):
...
```
[the remark this answers is not in the document]
Comment (JOhnsonKC201) [reply jcm/nudging.py:343]: Humidity follows the same contract as u, v and T here: a zero timescale is `0 * (ref - state)`, so a NaN in any reference propagates, and `winds_only` on dev already evaluates temperature that way. The jcm ERA5 loader clamps out-of-range levels and pads the seam, so it does not produce NaN references. Masking NaNs would change all four variables together, so I'd rather do that as its own PR than add it to this one only for humidity.
